datafusion/optimizer/src/pre_cast_lit_in_binary_comparison.rs

```diff
@@ -0,0 +1,295 @@
+            let right_type = right.get_type(schema);
+            // can't get the data type, just return the expr
+            if left_type.is_err() || right_type.is_err() {
+                return expr.clone();
+            }
+            let left_type = left_type.unwrap();
+            let right_type = right_type.unwrap();
+            if !left_type.eq(&right_type)
+                && is_support_data_type(&left_type)
+                && is_support_data_type(&right_type)
+                && is_comparison_op(op)
+            {
+                match (&left, &right) {
+                    (Expr::Literal(_), Expr::Literal(_)) => {
+                        // do nothing
+                    }
+                    (Expr::Literal(left_lit_value), _)
+                        if can_integer_literal_cast_to_type(
+                            left_lit_value,
+                            &right_type,
+                        ) =>
+                    {
+                        // cast the left literal to the right type
+                        return binary_expr(
+                            cast_to_other_scalar_expr(left_lit_value, &right_type),
+                            *op,
+                            right,
+                        );
+                    }
+                    (_, Expr::Literal(right_lit_value))
+                        if can_integer_literal_cast_to_type(
+                            right_lit_value,
+                            &left_type,
+                        ) =>
+                    {
+                        // cast the right literal to the left type
+                        return binary_expr(
+                            left,
+                            *op,
+                            cast_to_other_scalar_expr(right_lit_value, &left_type),
+                        );
+                    }
+                    (_, _) => {
+                        // do nothing
+                    }
+                };
+            }
+            // return the new binary op
+            binary_expr(left, *op, right)
+        }
+        // TODO: optimize in list
+        // Expr::InList { .. } => {}
+        // TODO: handle other expr type and dfs visit them
+        _ => expr,
+    }
+}
+
+fn cast_to_other_scalar_expr(origin_value: &ScalarValue, target_type: &DataType) -> Expr {
+    // null case
+    if origin_value.is_null() {
+        // if the origin value is null, just convert to another type of null value
+        // The target type must be satisfied `is_support_data_type` method, we can unwrap safely
+        return lit(ScalarValue::try_from(target_type).unwrap());
+    }
+    // no null case
+    let value: i64 = match origin_value {
+        ScalarValue::Int8(Some(v)) => *v as i64,
+        ScalarValue::Int16(Some(v)) => *v as i64,
+        ScalarValue::Int32(Some(v)) => *v as i64,
+        ScalarValue::Int64(Some(v)) => *v as i64,
+        other_type => {
+            panic!("Invalid type and value {:?}", other_type);
```

Comment:
I created a PR against this PR to remove these panics and use `DataFusionError::Internal` instead.

I quite often hit panics in the code that should not be possible, so I think we should use `Result` where possible.

Reply:
I check the data type before this method. 
The data type can hit thie panic which is guaranteed by `is_support_data_type`

```
fn is_support_data_type(data_type: &DataType) -> bool {
    // TODO support decimal with other data type
    matches!(
        data_type,
        DataType::Int8 | DataType::Int16 | DataType::Int32 | DataType::Int64
    )
}
```

Reply:
But I will change the result type to `Result<T>` and clean the code with panic.